### PR TITLE
GO-2125 add origin relation with objects input

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### Description

Add origin relation with objects with object source:
1. clipboard - files, which was created by pasting them using clipboard
2. drag and drop - files, created from drag and drop
3. web clipper - objects, created from web clipper
4. sharing extension - objects, created using sharing extension
5. import - objects, created using import
6. use cases - pre-installed objects

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Related Tickets & Documents
https://linear.app/anytype/issue/GO-2125/sourceorigin-relation-to-understand-the-input

### Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [x] 🙅 no documentation needed

### CODE REVIEW

File: pkg/lib/pb/model/protos/models.proto [hunk 1]
```diff
@@ -777,6 +777,17 @@ enum RelationFormat {
...
+enum ObjectOrigin {
+    user = 0;
+    clipboard = 1;
+    dragAndDrop = 2;
+    import = 3;
+    webClipper = 4;
+    sharingExtension = 5;
+    useCase = 6;
+    builtin = 7;
+}
```
Comment (ra3orblade) [inline pkg/lib/pb/model/protos/models.proto:789]: I would remove camelcase in webClipper and useCase words

File: core/indexer/indexer.go [hunk 2]
```diff
@@ -288,6 +294,28 @@ func (i *indexer) getObjectInfo(ctx context.Context, id string) (info smartblock
 	return
 }
 
+func (i *indexer) setFileOrigin(hash string, origin int) {
+	fileOrigin, err := i.fileStore.GetFileOrigin(hash)
```
Comment (deff7) [inline core/indexer/indexer.go:298]: This code should be inside `Details(ctx context.Context) (*types.Struct, error)` in
core/files/image.go
core/files/file.go

File: core/application/account_create.go [hunk 3]
```diff
@@ -112,7 +112,7 @@ func (s *Service) setAccountAndProfileDetails(ctx context.Context, req *pb.RpcAc
 		hash, err := bs.UploadFile(context.Background(), spaceID, pb.RpcFileUploadRequest{
 			LocalPath: req.GetAvatarLocalPath(),
 			Type:      model.BlockContentFile_Image,
-		})
+		}, model.ObjectOrigin_user)
```
Comment (deff7) [inline core/application/account_create.go:115]: Maybe create DTO for request and add origin there?

It should replicate fields of RpcFileUploadRequest plus contain Origin

Comment (AnastasiaShemyakinskaya) [reply core/application/account_create.go]: Done

File: core/indexer/indexer.go (hunk 2 above)
Comment (AnastasiaShemyakinskaya) [reply core/indexer/indexer.go]: Done

File: core/block/editor/file/file.go [hunk 4]
```diff
@@ -216,12 +221,12 @@ func (sf *sfile) DropFiles(req pb.RpcFileDropRequest) (err error) {
 	return
 }
 
-func (sf *sfile) UploadFileWithHash(blockId string, source FileSource) (UploadResult, error) {
+func (sf *sfile) UploadFileWithHash(blockID string, source FileSource, origin model.ObjectOrigin) (UploadResult, error) {
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/editor/file/file.go]: Done

File: core/block/editor/basic/extract_objects.go [hunk 5]
```diff
@@ -156,6 +156,7 @@ func createTargetObjectDetails(nameText string, layout model.ObjectTypeLayout) *
 		fields[bundle.RelationKeyName.String()] = pbtypes.String(nameText)
 	}
 
+	fields[bundle.RelationKeyOrigin.String()] = pbtypes.Int64(int64(model.ObjectOrigin_user))
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/editor/basic/extract_objects.go]: I added to CreateStateMigration template WithOrigin, which adds to pages origin user

File: pkg/lib/pb/model/protos/models.proto (hunk 1 above)
Comment (AnastasiaShemyakinskaya) [reply pkg/lib/pb/model/protos/models.proto:794]: Done

File: core/block/import/syncer/file.go [hunk 6]
```diff
@@ -44,7 +44,11 @@ func (fs *FileSyncer) Sync(id string, b simple.Block) error {
 			BlockId: b.Model().Id,
 		}
 	}
-	_, err := fs.service.UploadFileBlockWithHash(id, params)
+	dto := block.UploadRequest{
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/import/syncer/file.go:47]: AFAIU, logic with marked imported files is related to migrated files. And these files are already uploaded, so we mark them separately from uploader logic

File: core/block/import/objectcreator.go [hunk 7]
```diff
@@ -466,3 +469,23 @@ func (oc *ObjectCreator) setFileAsImported(st *state.State) {
 		}
 	}
 }
+
+func (oc *ObjectCreator) setFileOrigin(st *state.State, origin model.ObjectOrigin) {
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/import/objectcreator.go]: Fixed

File: util/objectorigin/helper.go [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package objectorigin
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply util/objectorigin/helper.go]: Removed it, as it's not necessary anymore

File: pkg/lib/pb/model/protos/models.proto [hunk 9]
```diff
@@ -782,6 +782,16 @@ enum RelationFormat {
 
 }
 
+enum ObjectOrigin {
+    clipboard = 0;
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply pkg/lib/pb/model/protos/models.proto]: I added in model none=0 for such cases

File: core/block/editor.go [hunk 10]
```diff
@@ -33,6 +33,21 @@ import (
 
 var ErrOptionUsedByOtherObjects = fmt.Errorf("option is used by other objects")
 
+type FileUploadRequest struct {
+	pb.RpcFileUploadRequest
+	Origin *model.ObjectOrigin
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/editor.go]: Yes, I added in model none=0 as I mentioned in previous comment

File: core/files/image.go [hunk 11]
```diff
@@ -39,6 +39,7 @@ type image struct {
 	spaceID         string
 	variantsByWidth map[int]*storage.FileInfo
 	service         *service
+	origin          int
```
Comment (deff7) [inline core/files/image.go:42]: maybe use model.origin type?

File: core/block/import/importer.go [hunk 12]
```diff
@@ -251,7 +254,7 @@ func (i *Import) ImportWeb(ctx context.Context, req *pb.RpcObjectImportRequest)
 	}
 
 	progress.SetProgressMessage("Create objects")
-	details, _ := i.createObjects(ctx, res, progress, req, allErrors)
+	details, _ := i.createObjects(ctx, res, progress, req, allErrors, 0)
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/import/importer.go]: Fixed

**Review by deff7 — APPROVED** [review]

**Pull request merged**